### PR TITLE
r?fzzzy ianbicking FxA integration

### PR DESCRIPTION
**Pull request by kitcambridge**
Whew, sorry this took so long! This is a first cut at adding FxA integration to PageShot. The "sign in" and "create account" buttons are wired up, and you can change your nickname. The styling could use some love. I'd love feedback on the React views, and the general approach.

Dev client credentials: `CLIENT_ID=209ee70cb3cdc851 CLIENT_SECRET=efde5874713abf2a8ec9d8af8488fae7c7e0a933dba08208f644fa93b5622dab` (We can check these into the repo; the FxA folks can provide us with real production credentials later).

### CODE REVIEW

File: addon/lib/main.js [hunk 1]
```diff
@@ -51,6 +52,15 @@ var shootPanel = panels.Panel({
 
 watchWorker(shootPanel);
 
+shootPanel.port.on("signIn", function () {
+  let oAuthHandler = new user.OAuthHandler('http://localhost:10080');
```
Comment (fzzzy) [inline addon/lib/main.js]: Don't forget to get this out of the environment before landing.

File: addon/lib/main.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ const helperworker = require("./helperworker");
 const { ToggleButton } = require('sdk/ui/button/toggle');
 const panels = require("sdk/panel");
 const { watchFunction, watchWorker } = require("./errors");
+const user = require("user");
```
Comment (ianb) [inline addon/lib/main.js]: Later on we'll want to rename this deviceid or something.  Otherwise we're all going to get really confused later.

File: addon/lib/main.js (hunk 1 above)
Comment (ianb) [inline addon/lib/main.js]: This can just be replaced with `backend`

File: server/src/server.js [hunk 3]
```diff
@@ -186,6 +196,132 @@ app.get("/:id/:domain", function (req, res) {
...
+          res.send({
+            access_token: accessTokenHex
+          });
+        });
+      });
+    });
+  }).catch(next);
+});
+
+function setState(deviceId, state) {
```
Comment (ianb) [inline server/src/server.js]: I'd like to keep SQL itself out of server.js.  Probably a new module to handle these?

File: server/src/users.js [hunk 4]
```diff
@@ -2,7 +2,7 @@ const db = require("./db");
 
 exports.checkLogin = function (userId, secret) {
   return db.select(
-    `SELECT secret FROM users WHERE id = $1`,
+    `SELECT secret FROM devices WHERE id = $1`,
```
Comment (ianb) [inline server/src/users.js]: oh, and you renamed the table, cool ;)

File: package.json [hunk 5]
```diff
@@ -8,6 +8,7 @@
   "dependencies": {
     "babel": "5.2.17",
     "body-parser": "1.12.4",
+    "boom": "^2.7.0",
```
Comment (ianb) [inline package.json]: We've been using strict versions in package.json (i.e., `"2.7.0"`)

File: server/src/dbschema.js [hunk 6]
```diff
@@ -2,16 +2,25 @@ const db = require("./db");
 const Keygrip = require('keygrip');
 
 const createSQL = `
-CREATE TABLE IF NOT EXISTS users (
+CREATE TABLE IF NOT EXISTS accounts (
+  id BYTEA PRIMARY KEY,
+  token BYTEA,
+  email varchar(256) NOT NULL
```
Comment (ianb) [inline server/src/dbschema.js]: Generally I try to use `TEXT` as often as possible

File: server/src/dbschema.js [hunk 7]
```diff
@@ -2,16 +2,25 @@ const db = require("./db");
 const Keygrip = require('keygrip');
 
 const createSQL = `
-CREATE TABLE IF NOT EXISTS users (
+CREATE TABLE IF NOT EXISTS accounts (
+  id BYTEA PRIMARY KEY,
+  token BYTEA,
```
Comment (ianb) [inline server/src/dbschema.js]: These are bytes?  That's kind of surprising I guess.

File: server/src/dbschema.js [hunk 8]
```diff
@@ -20,6 +29,13 @@ CREATE TABLE IF NOT EXISTS signing_keys (
   created TIMESTAMP DEFAULT NOW(),
   key TEXT
 );
+
+CREATE TABLE IF NOT EXISTS states (
+  state BYTEA PRIMARY KEY,
+  deviceid VARCHAR(200) REFERENCES devices(id) ON DELETE CASCADE
+);
+
+CREATE INDEX ON states(deviceid);
```
Comment (kitcambridge) [inline server/src/dbschema.js]: We'll want to give this index a name and follow up with #446, to avoid creating new indexes each time we restart the server.

File: server/src/errors.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+const boom = require('boom');
+
+exports.create = function create(statusCode, errno, message, data) {
```
Comment (ianb) [inline server/src/errors.js]: What's the advantage of boom here?

File: server/src/server.js [hunk 10]
```diff
@@ -186,6 +196,132 @@ app.get("/:id/:domain", function (req, res) {
   });
 });
 
+const oAuthBaseURI = 'http://127.0.0.1:9010/v1',
+  contentBaseURI = 'http://127.0.0.1:3030',
+  profileBaseURI = 'http://127.0.0.1:1111/v1',
+  oAuthClientId = 'ac7ee3c317531aab',
+  oAuthClientSecret = 'eb7c92bd4616a7c1f86595492f360f55cc453974b062e71456d2d4b104f307f8';
```
Comment (ianb) [inline server/src/server.js]: I guess these might cause problems being in the source?

Generally I think we should just read them from process.env.  We can make sure they are set properly on the live server.  I'm not sure what to do in development?

**Comment by kitcambridge** [comment]
I think this is finally ready!

File: server/src/server.js [hunk 11]
```diff
@@ -189,6 +215,76 @@ app.get("/:id/:domain", function (req, res) {
...
+      return registerAccount(req.userId, accountId, accessToken);
+    }).then(() => {
+      res.send({
+        access_token: accessToken
+      });
+    });
+  }).catch(next);
+});
+
+app.use(function (err, req, res, next) {
```
Comment (kitcambridge) [inline server/src/server.js]: I moved the error handler last so that `.catch(next)` can catch and report errors in promise chains.

File: addon/lib/main.js (hunk 2 above)
Comment (kitcambridge) [inline addon/lib/main.js]: Maybe the FxA additions belong in a separate `profile` module? Also, since the OAuth code lives elsewhere now, I should revert the changes to this file...

File: addon/lib/helperworker.js [hunk 12]
```diff
@@ -31,6 +50,64 @@ function resetPageMod(backend) {
...
+
+        let handler = new user.OAuthHandler(backend);
+        handler.getOAuthParams().then(defaults => {
+          worker.port.emit("account", id, toSuccessResponse());
+
+          let params = Object.assign({ action }, defaults);
+          return handler.logInWithParams(params).then(response => {
+            return handler.getProfileInfo();
+          }).then(profile => {
+            // TODO: Don't clobber profile info if the user has already set a
```
Comment (kitcambridge) [inline addon/lib/helperworker.js]: Maybe a separate `setDefaultProfileInfo` call that only updates your nickname and avatar if you haven't set one?

File: addon/data/viewerworker.js [hunk 13]
```diff
@@ -27,6 +27,51 @@ document.addEventListener("error", function (event) {
...
+  }));
+  document.dispatchEvent(event);
+});
+
+self.port.on("profileRefresh", function () {
+  var event = document.createEvent("CustomEvent");
+  event.initCustomEvent("profile-refresh", true, true, null);
+  document.dispatchEvent(event);
+});
+
```
Comment (ianb) [inline addon/data/viewerworker.js]: Maybe all this messaging could be simpler.  We basically want to keep data in sync between the page and the addon.  I think we can consider the addon the canonical source?  So we have messages from the page asking to do things: to sign in, to create an account, to set the display name, and to set the avatar.  From the addon we have: set the state.  That includes the login status, profile information, display name, etc.  The addon may have to wait to set the state if the page isn't ready to accept the message yet, though I think the way the helper initializes, all the Javascript on the page should have executed (including event listeners) before the worker starts up.

So it seems to me we should have some events from the page: request-signin, request-create, set-profile-info

Then from the addon to the page: set-profile-info.

I've used a pattern of names-with-dashes in event names, and camelCase in message names.  This is a silly.  For the most part viewerworker just pipes stuff through, we should probably keep the exact same names, and potentially do some of these in a loop.

Comment (kitcambridge) [inline addon/data/viewerworker.js]: Agreed on simplifying.

* We can ax `got-profile` and `got-profile-update`, since they're taken care of by `profile-refresh` to set the state. If you're already signed in, but the page didn't pick up on that, we can just emit `profile-refresh`.
* Also, `got-account` just tells the page, "Hey, I'm starting the OAuth flow." I thought this might surface errors in the OAuth flow, too, but it doesn't, and they're not user-actionable. So we can ax that.

File: addon/lib/user.js [hunk 14]
```diff
@@ -1,7 +1,11 @@
+const { Cc, Ci, Cu } = require('chrome');
```
Comment (ianb) [inline addon/lib/user.js]: I think jshint gives unused import warnings for some of these.  Not sure how I feel – this is a common import, but I also like to keep things jshint clean even when it makes me do things that feel nitpicky.

File: addon/lib/user.js [hunk 15]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
 exports.getUserInfo = function () {
   return ss.storage.userInfo;
 };
+
+exports.updateProfileInfo = function (profile) {
```
Comment (ianb) [inline addon/lib/user.js]: I guess `ss.storage.profileInfo` is info taken from your FxA, but `ss.storage.userInfo` is actually device info.  Maybe the latter should be renamed to clarify?  As it is even I get confused about which is which.

File: addon/lib/user.js [hunk 16]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
...
+  if (! profile) {
```
Comment (ianb) [inline addon/lib/user.js]: Why would you get an empty profile?  I might be include to throw an error here.

File: addon/lib/user.js [hunk 17]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
...
+    return;
+  }
+  let currentInfo = ss.storage.profileInfo;
```
Comment (ianb) [inline addon/lib/user.js]: since we have to set ss.storage.profileInfo anyway, this could be `let currentInfo = ss.storage.profileInfo || {}` and forget the next if statement

File: addon/lib/user.js [hunk 18]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
...
+  if (! currentInfo) {
+    currentInfo = ss.storage.profileInfo = {};
+  }
+  let attrs = Object.keys(profile);
+  for (let attr of attrs) {
```
Comment (ianb) [inline addon/lib/user.js]: This logic is complicated...?  Could it just be:

```javascript
for (let attr in profile) {
  currentInfo[attr] = profile[attr];
}
```

?  Or are you worried about null values that don't mean erase, or...?

File: addon/lib/user.js [hunk 19]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
...
+    }
+  }
+  ss.storage.profileInfo = currentInfo;
+};
+
+exports.getProfileInfo = function () {
+  return ss.storage.profileInfo;
+};
+
+exports.updateLogin = function (backend, info) {
```
Comment (ianb) [inline addon/lib/user.js]: I think this more means `updateProfile` or `updateAccount`?  This is basically updating the `accounts` table with profile info for the user, right?  I suppose it is also assigning the reference between the device and the account...?

File: addon/lib/user.js [hunk 20]
```diff
@@ -64,3 +68,129 @@ function makeUuid() {
...
+    }).then(response => {
+      this.profileDeferred.resolve(new FxAccountsProfileClient({
+        serverURL: parameters.profile_uri,
+        token: response.access_token
+      }));
+      return response;
+    });
+  }
+
+  logIn() {
```
Comment (ianb) [inline addon/lib/user.js]: Is this method used anywhere?

File: server/src/clientglue.js [hunk 21]
```diff
@@ -21,9 +22,7 @@ exports.setModel = function (data) {
 
 function render() {
   setGitRevision(model.gitRevision);
-  let attrs = {
-    staticLink: staticLink
-  };
+  let attrs = { staticLink, ready, signUp, signIn, getProfile, updateProfile };
```
Comment (ianb) [inline server/src/clientglue.js]: I don't think these need to be properties?  staticLink is a special case because it is used on the server and is request-specific.  But the rest are only used on the client and can be called globally.  So frame.js can just include these functions and call them.

File: server/src/views/profile.js [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+function delay(duration) {
+  return new Promise(resolve => setTimeout(resolve, duration));
+}
+
+exports.ProfileButton = class ProfileButton extends React.Component {
+  constructor(props) {
+    super(props);
+    this.onDocumentClick = this.dismiss.bind(this);
+    this.onRefreshProfileBound = this.onRefreshProfile.bind(this);
+    this.state = {
```
Comment (ianb) [inline server/src/views/profile.js]: avatar, nickname, and email can probably all be props, while isExpanded and isFetching are probably right to have as state.  I think `getInitialState` is usually how that state is set

Comment (fzzzy) [inline server/src/views/profile.js]: getInitialState was the way to do it with old style classes, but with es6 classes the constructor is the recommended place for it

File: server/src/views/profile.js [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+    this.props.updateProfile(profile).then(() => {
+      this.fetchProfile();
+    }).catch(err => {
+      console.error("Error updating profile", err);
+    });
+  }
+
+  componentDidMount() {
+    window.addEventListener("click", this.onDocumentClick);
+    document.addEventListener("profile-refresh", this.onRefreshProfileBound);
```
Comment (ianb) [inline server/src/views/profile.js]: If that data is props instead of state, I expect the handler will be global and will trigger a refresh

**Comment by ianb** [comment]
That's it from me!  We'll have @brampitoyo look at the visual design more closely after merge.

File: addon/lib/user.js (hunk 18 above)
Comment (kitcambridge) [inline addon/lib/user.js]: > Or are you worried about null values that don't mean erase...

Good catch! I overloaded `updateProfileInfo` to mean, "update this profile record with the user's changes," and "fill in default values from FxA if the user hasn't provided them yet," because I didn't know how I wanted them to fit together. This definitely calls for separate methods now, though.

File: server/src/views/profile.js (hunk 22 above)
Comment (kitcambridge) [inline server/src/views/profile.js]: I could make them props, and have `Frame` pass a function that updates the profile. In that case, though, wouldn't `avatar`, `nickname`, and `email` still need to be state (just on `Frame`, not `ProfileButton`)?

Comment (ianb) [inline server/src/views/profile.js]: In clientglue.js I think it should keep the profile information in the model that it uses for render().  Then clientglue is what catches the profile-update event from viewerworker, and then rerenders the entire page.

Comment (ianb) [inline server/src/views/profile.js]: clientglue is kind of the controller for the entire shot page
